test.py

```diff
@@ -1,124 +1,527 @@
+        with self.assertRaises(NotImplementedError):
+            self.api_provider.get_diff("https://github.com/servo/servo/pull/1234.diff")
+
+
+    def test_set_assignee(self):
+        with self.assertRaises(NotImplementedError):
+            self.api_provider.set_assignee("jdm", 1234)
+
+
+# TODO - add tests for exception handling
+class TestGithubApiProvider(unittest.TestCase):
+    class FakeHeader():
+            def get(self, something):
+                pass
+
+    def setUp(self):
+        self.owner = "servo"
+        self.repo = "servo"
+        self.gh_provider = GithubApiProvider("jdm", "a453b3923e893f0383cd2893f...", self.owner, self.repo)
+        self.gh_provider.api_req = Mock()
+
+    @patch('githubapiprovider.gzip')
+    @patch('githubapiprovider.urllib2')
+    @patch('githubapiprovider.base64')
+    def test_api_req(self, base64_mock, urllib_mock, gzip_mock):
+        class FakeRequest():
+            def __init__(self):
+                self.headers = {}
+                self.get_method = lambda: "Get"
+
+
+            def add_header(self, key, value):
+                self.headers[key] = value
+
+
+        class FakeHeader():
+            def get(self, type):
+                return 'gzip'
+
+
+        class FakeResponse():
+            def __init__(self, fakeHeader):
+                self.header = fakeHeader
+
+
+            def info(self):
+                return self.header
+
+
+            def read(self):
+                return "content"
+
+        faux_request = FakeRequest()
+        base64_mock.standard_b64encode = Mock(return_value="User:Token")
+        urllib_mock.Request = Mock(return_value=faux_request)
+        urllib_mock.urlopen = Mock(return_value=FakeResponse(FakeHeader()))
+        gzip_mock.GzipFile = Mock(return_value=FakeResponse(FakeHeader()))
+        gh_provider = GithubApiProvider("jdm", "a453b3923e893f0383cd2893f...", self.owner, self.repo)
+        gh_provider.api_req("GET", "https://api.github.com/repos/servo/servo/contributors", data={"test":"data"}, media_type="Test Media")
+        urllib_mock.Request.assert_called_with("https://api.github.com/repos/servo/servo/contributors", json.dumps({"test":"data"}), {'Content-Type':'application/json'})
+        urllib_mock.urlopen.assert_called_with(faux_request)
+
+
+    def test_parse_header_links(self):
+        header_links =  '<https://api.github.com/repos/servo/servo/contributors?page=2>; rel="next", <https://api.github.com/repos/servo/servo/contributors?page=11>; rel="last"'
+        expected = {'last': 'https://api.github.com/repos/servo/servo/contributors?page=11', 'next': 'https://api.github.com/repos/servo/servo/contributors?page=2'}
+        gh_provider = GithubApiProvider("jdm", "a453b3923e893f0383cd2893f...", self.owner, self.repo)
+        self.assertEquals(expected, gh_provider.parse_header_links(header_links))
+        self.assertEquals(None, gh_provider.parse_header_links(None))
+
+
+    def test_is_new_contributor(self):
```

Comment:
I like the idea of the test (ie. verifying that a contributor is classified as new or not new appropriately), but in practice I find the details incredibly difficult to read and I'm not certain how to improve the situation :/